File: app/src/main/java/de/kuschku/util/ui/preferences/SeekBarPreference.java
/*
 * QuasselDroid - Quassel client for Android
 * Copyright (C) 2016 Janne Koschinski
 * Copyright (C) 2016 Ken Børge Viktil
 * Copyright (C) 2016 Magnus Fjell
 * Copyright (C) 2016 Martin Sandsmark <martin.sandsmark@kde.org>
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package de.kuschku.util.ui.preferences;

import android.content.Context;
import android.content.res.TypedArray;
import android.preference.Preference;
import android.support.v7.widget.AppCompatSeekBar;
import android.util.AttributeSet;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.SeekBar;
import android.widget.SeekBar.OnSeekBarChangeListener;
import android.widget.TextView;

import de.kuschku.quasseldroid_ng.R;

public class SeekBarPreference extends Preference implements OnSeekBarChangeListener {

    private static final String ANDROIDNS = "http://schemas.android.com/apk/res/android";
    private static final String APPLICATIONNS = "http://robobunny.com";
    private static final int DEFAULT_VALUE = 50;
    private final String TAG = getClass().getName();
    private int mMaxValue = 100;
    private int mMinValue = 0;
    private int mInterval = 1;
    private int mCurrentValue;
    private AppCompatSeekBar mSeekBar;

    private TextView mStatusText;

    public SeekBarPreference(Context context, AttributeSet attrs) {
        super(context, attrs);
        initPreference(context, attrs);
    }

    public SeekBarPreference(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        initPreference(context, attrs);
    }

    private void initPreference(Context context, AttributeSet attrs) {
        setValuesFromXml(attrs);
        mSeekBar = new AppCompatSeekBar(context, attrs);
        mSeekBar.setMax(mMaxValue - mMinValue);
        mSeekBar.setOnSeekBarChangeListener(this);

        setWidgetLayoutResource(R.layout.seek_bar_preference);
    }

    private void setValuesFromXml(AttributeSet attrs) {
        mMaxValue = attrs.getAttributeIntValue(ANDROIDNS, "max", 100);
        mMinValue = attrs.getAttributeIntValue(APPLICATIONNS, "min", 0);

        try {
            String newInterval = attrs.getAttributeValue(APPLICATIONNS, "interval");
            if (newInterval != null)
                mInterval = Integer.parseInt(newInterval);
        } catch (Exception e) {
            Log.e(TAG, "Invalid interval value", e);
        }

    }

    @Override
    protected View onCreateView(ViewGroup parent) {
        View view = super.onCreateView(parent);

        // The basic preference layout puts the widget frame to the right of the title and summary,
        // so we need to change it a bit - the seekbar should be under them.
        LinearLayout layout = (LinearLayout) view;
        layout.setOrientation(LinearLayout.VERTICAL);

        return view;
    }

    @Override
    public void onBindView(View view) {
        super.onBindView(view);
        if (view != null) {
            mSeekBar = (AppCompatSeekBar) view.findViewById(R.id.seekBarPrefSeekBar);
            mSeekBar.setMax(mMaxValue - mMinValue);
            mSeekBar.setOnSeekBarChangeListener(this);
        }

        updateView(view);
    }

    /**
     * Update a SeekBarPreference view with our current state
     *
     * @param view
     */
    protected void updateView(View view) {

        try {
            mStatusText = (TextView) view.findViewById(R.id.seekBarPrefValue);

            mStatusText.setText(String.valueOf(mCurrentValue));
            mStatusText.setMinimumWidth(30);

            mSeekBar.setProgress(mCurrentValue - mMinValue);

        } catch (Exception e) {
            Log.e(TAG, "Error updating seek bar preference", e);
        }

    }

    @Override
    public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
        int newValue = progress + mMinValue;

        if (newValue > mMaxValue)
            newValue = mMaxValue;
        else if (newValue < mMinValue)
            newValue = mMinValue;
        else if (mInterval != 1 && newValue % mInterval != 0)
            newValue = Math.round(((float) newValue) / mInterval) * mInterval;

        // change rejected, revert to the previous value
        if (!callChangeListener(newValue)) {
            seekBar.setProgress(mCurrentValue - mMinValue);
            return;
        }

        // change accepted, store it
        mCurrentValue = newValue;
        mStatusText.setText(String.valueOf(newValue));
        persistInt(newValue);

    }

    @Override
    public void onStartTrackingTouch(SeekBar seekBar) {
    }

    @Override
    public void onStopTrackingTouch(SeekBar seekBar) {
        notifyChanged();
    }


    @Override
    protected Object onGetDefaultValue(TypedArray ta, int index) {
        return ta.getInt(index, DEFAULT_VALUE);

    }

    @Override
    protected void onSetInitialValue(boolean restoreValue, Object defaultValue) {

        if (restoreValue) {
            mCurrentValue = getPersistedInt(mCurrentValue);
        } else {
            int temp = 0;
            try {
                temp = (Integer) defaultValue;
            } catch (Exception ex) {
                Log.e(TAG, "Invalid default value: " + defaultValue.toString());
            }

            persistInt(temp);
            mCurrentValue = temp;
        }

    }

    /**
     * make sure that the seekbar is disabled if the preference is disabled
     */
    @Override
    public void setEnabled(boolean enabled) {
        super.setEnabled(enabled);
        mSeekBar.setEnabled(enabled);
    }

    @Override
    public void onDependencyChanged(Preference dependency, boolean disableDependent) {
        super.onDependencyChanged(dependency, disableDependent);

        //Disable movement of seek bar when dependency is false
        if (mSeekBar != null) {
            mSeekBar.setEnabled(!disableDependent);
        }
    }
}

File: app/src/main/java/com/robobunny/SeekBarPreference.kt
package com.robobunny

import android.content.Context
import android.content.res.TypedArray
import android.support.v7.preference.Preference
import android.support.v7.preference.PreferenceViewHolder
import android.support.v7.widget.AppCompatSeekBar
import android.util.AttributeSet
import android.util.Log
import android.widget.SeekBar
import android.widget.TextView
import butterknife.BindView
import butterknife.ButterKnife
import de.kuschku.quasseldroid_ng.R

/*
 * Copyright (c) 2015 IRCCloud, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
class SeekBarPreference : Preference, SeekBar.OnSeekBarChangeListener {

  private val TAG = javaClass.name

  private var maxValue = 100
  private var minValue = 0
  private var interval = 1
  private var currentValue: Int = 0

  private var unitsLeftText = ""
  private var unitsRightText = ""

  @BindView(R.id.seekBarPrefSeekBar)
  @JvmField
  var seekBar: AppCompatSeekBar? = null

  @BindView(R.id.seekBarPrefValue)
  @JvmField
  var statusText: TextView? = null

  @BindView(R.id.seekBarPrefUnitsLeft)
  @JvmField
  var unitsLeft: TextView? = null

  @BindView(R.id.seekBarPrefUnitsRight)
  @JvmField
  var unitsRight: TextView? = null

  constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) :
    super(context, attrs, defStyleAttr) {
    initPreference(context, attrs)
  }

  constructor(context: Context, attrs: AttributeSet) :
    super(context, attrs) {
    initPreference(context, attrs)
  }

  private fun initPreference(context: Context, attrs: AttributeSet) {
    setValuesFromXml(attrs)
    layoutResource = R.layout.preference_vertical
    widgetLayoutResource = R.layout.preference_seekbar
  }

  private fun setValuesFromXml(attrs: AttributeSet) {
    maxValue = attrs.getAttributeIntValue(NAMESPACE_ANDROID, "max", 100)
    minValue = attrs.getAttributeIntValue(NAMESPACE_ROBOBUNNY, "min", 0)
    unitsLeftText = getAttributeStringValue(attrs, NAMESPACE_ROBOBUNNY, "unitsLeft", "")
    val units = getAttributeStringValue(attrs, NAMESPACE_ROBOBUNNY, "units", "")
    unitsRightText = getAttributeStringValue(attrs, NAMESPACE_ROBOBUNNY, "unitsRight", units)
    try {
      val newInterval = attrs.getAttributeValue(NAMESPACE_ROBOBUNNY, "interval")
      if (newInterval != null)
        interval = Integer.parseInt(newInterval)
    } catch (e: Exception) {
      Log.e(TAG, "Invalid interval value", e)
    }
  }

  private fun getAttributeStringValue(attrs: AttributeSet, namespace: String, name: String,
                                      defaultValue: String) =
    attrs.getAttributeValue(namespace, name) ?: defaultValue

  override fun onBindViewHolder(holder: PreferenceViewHolder?) {
    super.onBindViewHolder(holder)
    holder?.itemView?.let { view ->
      ButterKnife.bind(this, view)
      seekBar?.max = maxValue - minValue
      seekBar?.setOnSeekBarChangeListener(this)
      statusText?.text = currentValue.toString()
      statusText?.minimumWidth = 30
      seekBar?.progress = currentValue - minValue
      unitsRight?.text = this.unitsRightText
      unitsLeft?.text = this.unitsLeftText
    }
  }

  override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
    var newValue = progress + minValue
    if (newValue > maxValue)
      newValue = maxValue
    else if (newValue < minValue)
      newValue = minValue
    else if (interval != 1 && newValue % interval != 0)
      newValue = Math.round(newValue.toFloat() / interval) * interval
    // change rejected, revert to the previous value
    if (!callChangeListener(newValue)) {
      seekBar.progress = currentValue - minValue
      return
    }
    // change accepted, store it
    currentValue = newValue
    statusText?.text = newValue.toString()
    persistInt(newValue)
  }

  override fun onStartTrackingTouch(seekBar: SeekBar) = Unit
  override fun onStopTrackingTouch(seekBar: SeekBar) = notifyChanged()
  override fun onGetDefaultValue(ta: TypedArray, index: Int) = ta.getInt(index, DEFAULT_VALUE)
  override fun onSetInitialValue(restoreValue: Boolean, defaultValue: Any?) {
    if (restoreValue) {
      currentValue = getPersistedInt(currentValue)
    } else {
      var temp = 0
      try {
        temp = defaultValue as Int
      } catch (ex: Exception) {
        Log.e(TAG, "Invalid default value: " + defaultValue.toString())
      }
      persistInt(temp)
      currentValue = temp
    }
  }

  /**
   * make sure that the seekbar is disabled if the preference is disabled
   */
  override fun setEnabled(enabled: Boolean) {
    super.setEnabled(enabled)
    seekBar?.isEnabled = enabled
  }

  override fun onDependencyChanged(dependency: Preference, disableDependent: Boolean) {
    super.onDependencyChanged(dependency, disableDependent)
    //Disable movement of seek bar when dependency is false
    seekBar?.isEnabled = !disableDependent
  }

  companion object {
    private const val NAMESPACE_ANDROID = "http://schemas.android.com/apk/res/android"
    private const val NAMESPACE_ROBOBUNNY = "http://robobunny.com"
    private const val DEFAULT_VALUE = 50
  }
}